Recursively compute variables

spencerkclark opened the pull request:
Closes #3 

This allows one to compose variable objects whose functions take other computed variables as arguments.  For example:
```python
from aospy import Var

a = Var(name='a')
b = Var(name='b', variables=(a, ), func=lambda x: 2. * x)
c = Var(name='c', variables=(a, b), func=lambda x, y: x + y)
```

Commit:
Recursive compute variables

Commit:
Add what's new

Commit:
Merge branch 'develop' into recursive-var-compute

spencerkclark commented:
Note of course this doesn't deal with saving intermediates to disk (or checking for saved intermediates), which is also discussed in #3, but I think what's here is already a reasonable improvement.

spencerahill commented:
Wow, this is surprisingly elegant!  The logic and tests look good to me.

One big thing: have we tested this out within a `Calc` yet?  I expect that, given these unit tests, everything will be fine, but I still think it would be good to have an explicit integration test for this.

Besides that, a couple documentation requests:

- Can you give `recursively_compute_variable` a full docstring?
- More substantively, can we highlight this functionality in the docs somewhere?  In scanning through our existing pages, I'm not sure where this fits best.  Perhaps [here](https://aospy.readthedocs.io/en/latest/examples.html#physical-variables) within the Examples page.

Commit:
Add test for recursive calculation; fix minor issue

spencerkclark commented on aospy/utils/io.py:
```diff
@@ -154,9 +154,14 @@ def data_name_gfdl(name, domain, data_type, intvl_type, data_yr,
-        if isinstance(files_list, str):
-            files_list = [files_list]
-        subprocess.call(['dmget'] + files_list)
+        subprocess.call(['dmget'] + archive_files)
```
I was finding that the test suite would hang if `dmget` was called on an existing file that was not stored in `/archive`.  I was able to reproduce this outside of aospy, so it has nothing to do with the changes I made to allow for the recursive calculation of variables; they must have made some updates to `dmget`.  The logic here filters out files that aren't stored on archive before passing them to `dmget`.

spencerkclark commented on aospy/data_loader.py:
```diff
@@ -296,7 +296,7 @@ def recursively_compute_variable(self, var, start_date=None, end_date=None,
                 v, start_date, end_date, time_offset, **DataAttrs)
                     for v in var.variables]
-            return var.func(*data)
+            return var.func(*data).rename(var.name)
```
Indeed, good suggestion to add a test for this.  One minor change and things work smoothly!

spencerkclark reviewed (commented):
Thanks @spencerahill!  I agree this should be highlighted in the documentation.  It will certainly simplify my object library quite a bit.  I'll add a full docstring to `recursively_compute_variable` and an example to the official docs sometime later today or tomorrow.

spencerahill commented:
All sounds/looks good.  Just ping me when you're ready for final review.

Commit:
Add documentation for this new feature

- Add a full docstring for compute_variable_recursively
- Add an example of this functionality to the official documentation
- Make test more succinct in test_calc_basic

Commit:
Further streamline test logic

spencerkclark commented:
I guess I spoke too soon in https://github.com/spencerahill/aospy/issues/259; that's the failing test here :)

@spencerahill anyway, whenever you get a chance, this should be ready for a final review!

spencerahill commented:
>I guess I spoke too soon in #259; that's the failing test here :)

Bummer.  Still, it seems like the frequency of the failures has gone down...maybe it's on its way toward going away completely.

In it goes!  Thanks @spencerkclark, this is great.
